zero/src/rpc/jerigon.rs

```diff
@@ -53,9 +116,75 @@ where
                 compact: hex::decode(block_witness.strip_prefix("0x").unwrap_or(&block_witness))
                     .context("invalid hex returned from call to eth_getWitness")?,
             }),
-            txn_info: tx_results.into_iter().map(|it| it.result).collect(),
+            txn_info,
             code_db: Default::default(),
         },
         other_data,
     })
 }
+
+/// Processes the transactions in the given block and updates the code db.
+pub async fn process_transactions<ProviderT, TransportT>(
+    block: &Block,
+    provider: &ProviderT,
+    tx_traces: &[&BTreeMap<__compat_primitive_types::H160, TxnTrace>],
+) -> anyhow::Result<Vec<Option<JumpDestTableWitness>>>
+where
+    ProviderT: Provider<TransportT>,
+    TransportT: Transport + Clone,
+{
+    let futures = block
+        .transactions
+        .as_transactions()
+        .context("No transactions in block")?
+        .iter()
+        .zip(tx_traces)
+        .map(|(tx, &tx_trace)| process_transaction(provider, tx, tx_trace))
+        .collect::<FuturesOrdered<_>>();
+
+    let vec_of_res = futures.collect::<Vec<_>>().await;
+    vec_of_res.into_iter().collect::<Result<Vec<_>, _>>()
+}
+
+/// Processes the transaction with the given transaction hash and updates the
+/// accounts state.
+pub async fn process_transaction<ProviderT, TransportT>(
```

Comment:
Same question about duplicated code. Could we unify the logic (e.g. if `tx_trace` is passed use if, if not go fetch it).

Reply:
Valid point, let's keep the discussion in comment above.